File: semana_7_patrones_comportamiento/codigo_ejemplos/ejemplo2_strategy.py
"""
EJEMPLO 2: Strategy en Python
(Version cruzada del ejemplo C++ del PDF - Semana 7)
"""


class Estrategia:
    def ejecutar(self):
        pass


class EstrategiaA(Estrategia):
    def ejecutar(self):
        print("Algoritmo A")


class EstrategiaB(Estrategia):
    def ejecutar(self):
        print("Algoritmo B")


class Contexto:
    def __init__(self, estrategia):
        self.estrategia = estrategia
    
    def set_estrategia(self, estrategia):
        self.estrategia = estrategia
    
    def operar(self):
        self.estrategia.ejecutar()


a = EstrategiaA()
b = EstrategiaB()
ctx = Contexto(a)
ctx.operar()
ctx.set_estrategia(b)
ctx.operar()

File: semana_7_patrones_comportamiento/codigo_ejemplos/ejemplo2_strategy.cpp
/*
 * EJEMPLO 2: Strategy en C++
 * (Ejemplo del PDF - Semana 7)
 */

#include <iostream>

using namespace std;

class Estrategia {
public:
    virtual void ejecutar() = 0;
};

class EstrategiaA : public Estrategia {
public:
    void ejecutar() override { 
        cout << "Algoritmo A\n"; 
    }
};

class EstrategiaB : public Estrategia {
public:
    void ejecutar() override { 
        cout << "Algoritmo B\n"; 
    }
};

class Contexto {
    Estrategia* estrategia;
public:
    Contexto(Estrategia* e) : estrategia(e) {}
    void setEstrategia(Estrategia* e) { estrategia = e; }
    void operar() { estrategia->ejecutar(); }
};

int main() {
    EstrategiaA a;
    EstrategiaB b;
    Contexto ctx(&a);
    ctx.operar();
    ctx.setEstrategia(&b);
    ctx.operar();
}
